packages/xod-client/src/user/actions.js

```diff
@@ -51,6 +34,69 @@ export const fetchGrant = () => dispatch =>
       )
     );
 
+/**
+ * Function that sends some request using Authorization token
+ * and tries to refresh the token on error 403.
+ * If it still fails it returns rejected Promise with an error.
+ *
+ * It accepts one argument of Function type with the signature:
+ * (Headers -> Promise FetchResult FetchError)
+ *
+ * Where `Headers` is an object, which will be modified with
+ * `Authorization` header.
+ */
+const requestAuthorized = (requestAction, headers) => (dispatch, getState) => {
+  const headersWithAuthorization = R.compose(
+    foldMaybe(headers, accessToken => ({
+      Authorization: `Bearer ${accessToken}`,
+      ...headers,
+    })),
+    getAccessToken
+  )(getState());
+
+  const run = () => requestAction(headersWithAuthorization);
+  return run()
+    .then(res => {
+      if (res.status > 400) {
+        return dispatch(refreshGrant()).then(() => run());
+      }
+      return res;
+    })
+    .then(res => {
+      if (!res.ok) {
+        const err = new Error(res.statusText);
+        err.status = res.status;
+        throw err;
+      }
+      return res.json();
+    })
+    .catch(() => {
+      dispatch(addError(Messages.SERVICE_UNAVAILABLE));
+    });
+};
+
+const fetchLimits = headers => fetch(getCompileLimitUrl(), { headers });
+
+export const updateCompileLimit = (startup = false) => dispatch => {
+  const basicHeaders = startup ? { 'x-launch': 'true' } : {};
+  return dispatch(requestAuthorized(fetchLimits, basicHeaders))
+    .catch(() => null)
+    .then(limit =>
+      dispatch({
+        type: ActionTypes.UPDATE_COMPILE_LIMIT,
+        payload: limit,
```

Comment:
`Compilations: undefined`?  :thinking:

Reply:
...or `currently offline`?

Reply:
The second one :)

Reply:
Okay, I won't update the limits if Promise rejected with an error. ;)